### PR TITLE
Exclusion edge case issue 1404

### PR DESCRIPTION
**Pull request by RidhamShah**
This PR addresses edge cases related to the inclusion and exclusion of Users and Groups already enrolled in an experiment when either a user or a group is added to the exclusion list based on the Assignment Unit and Consistency Rule.

Cases Covered:

- Group added to the exclusion list:
  - Experiment on Individual Consistency and Individual Assignment Unit:
    - Already enrolled users are not removed.
    - New users will be excluded by the exclusion logic.
  - Experiment on Group Assignment Unit and Individual Consistency:
    - Already enrolled users are not removed.
    - Group enrollment data is removed.
    - New users will be excluded by the exclusion logic.
  - Experiment on Group Assignment Unit and Group Consistency:
    - Already marked users are removed.
    - New users from the group will be excluded by the exclusion logic.
- User added to the exclusion list:
  - Experiment on Individual Consistency and Individual Assignment Unit:
    - The user is removed.
    - New users will not be excluded.
  - Experiment on Group Assignment Unit and Individual Consistency:
    - The user is excluded.
    - Group enrollment data is removed.
  - Experiment on Group Assignment Unit and Group Consistency:
    - The user is excluded.
    - Group data is removed and the group is excluded.

### CODE REVIEW

File: backend/packages/Upgrade/src/api/services/SegmentService.ts [hunk 1]
```diff
@@ -258,6 +285,11 @@ export class SegmentService {
     return queryBuilder;
   }
 
+  public async getExclusionSegmentUpdateData(segmentId: string) {
```
[the remark this answers is not in the document]
Comment (RidhamShah) [reply backend/packages/Upgrade/src/api/services/SegmentService.ts]: renamed

File: backend/packages/Upgrade/src/api/services/SegmentService.ts [hunk 2]
```diff
@@ -620,4 +670,123 @@ export class SegmentService {
...
+            }
+          }
+        }
+        if (newUsers.length) {
+          // Case 1: User already visited
+          const excludedUsers = await this.individualEnrollmentRepository.find({
+            where: { experiment: { id: experiment.id }, user: In(newUsers) },
+          });
+
+          const excludedUsersGroups = excludedUsers
```
[the remark this answers is not in the document]
Comment (RidhamShah) [reply backend/packages/Upgrade/src/api/services/SegmentService.ts]: code updated

File: backend/packages/Upgrade/src/api/services/ExperimentAssignmentService.ts [hunk 3]
```diff
@@ -1690,18 +1690,32 @@ export class ExperimentAssignmentService {
   ): Promise<[Experiment[], { experiment: Experiment; reason: string; matchedGroup: boolean }[]]> {
     const segmentObj = {};
 
+    let includedExperiments: Experiment[] = [];
+    let excludedExperiments = [];
+
```
[the remark this answers is not in the document]
Comment (RidhamShah) [reply backend/packages/Upgrade/src/api/services/ExperimentAssignmentService.ts]: code updated

File: backend/packages/Upgrade/src/api/services/ExperimentAssignmentService.ts [hunk 4]
```diff
@@ -1690,18 +1690,32 @@ export class ExperimentAssignmentService {
...
+    const experimentIds = experiments
```
[the remark this answers is not in the document]
Comment (RidhamShah) [reply backend/packages/Upgrade/src/api/services/ExperimentAssignmentService.ts]: code updated

File: backend/packages/Upgrade/src/api/services/SegmentService.ts [hunk 5]
```diff
@@ -620,4 +670,123 @@ export class SegmentService {
...
+              Omit<IndividualExclusion, 'id' | 'createdAt' | 'updatedAt' | 'versionNumber'>
+            > = enrolledUsers.map((user) => {
+              return {
+                user,
+                experiment,
+                groupId: user?.workingGroup?.[experiment.group],
+                exclusionCode: EXCLUSION_CODE.EXCLUDED_DUE_TO_GROUP_LOGIC,
+              };
+            });
+            await this.individualExclusionRepository.saveRawJson(individualExclusionDocs);
```
[the remark this answers is not in the document]
Comment (RidhamShah) [reply backend/packages/Upgrade/src/api/services/SegmentService.ts]: code updated

File: backend/packages/Upgrade/src/api/repositories/ExperimentSegmentExclusionRepository.ts [hunk 6]
```diff
@@ -49,6 +49,30 @@ export class ExperimentSegmentExclusionRepository extends Repository<ExperimentS
       });
   }
 
+  public async getExclusionSegmentUpdateDoc(segmentId: string): Promise<Partial<ExperimentSegmentExclusion[]>> {
```
[the remark this answers is not in the document]
Comment (RidhamShah) [reply backend/packages/Upgrade/src/api/repositories/ExperimentSegmentExclusionRepository.ts]: renamed

**Review by VivekFitkariwala — APPROVED** [review]

**Comment by danoswaltCL** [comment]
Is this ready? Vivek made comments and then approved, I don't know what that means.

**Pull request merged**